### PR TITLE
DIM Sync

### PR DESCRIPTION
**Pull request by bhollis**
This implements [DIM Sync](https://github.com/DestinyItemManager/DIM/wiki/DIM-Sync-(new-storage-for-tags,-loadouts,-and-settings)), the client part of the new DIM API. Read that wiki for more information, but suffice it to say this is a big change to how we handle storage!

This is currently opt-in - it's a default-checked option when you log in, and already-logged-in users will get a prompt asking whether they want it (and when they do, we'll automatically download a backup file). Either way, we'll write to both the old and the new storage in parallel in hopes of keeping things in sync just in case. Turning off DIM Sync goes back to the old Redux storage and IDB/Google Drive storage, while when it's on we have a new Redux structure and new IDB storage.

There's a lot that went into this. It's a huge PR to look at. I think what I'd like to do is try to get this into Beta ASAP and let people find bugs. Does that sound good? Also, I'm planning on bumping DIM to v6.0 as part of this release.

image

Screen Shot 2020-03-15 at 10 23 31 PM
Screen Shot 2020-03-15 at 10 34 59 PM

image

### CODE REVIEW

File: src/app/utils/util.ts [hunk 1]
```diff
@@ -58,3 +58,29 @@ export function dedupePromise<T extends any[], K>(
 export function delay(ms: number) {
   return new Promise((resolve) => setTimeout(resolve, ms));
 }
+
+/** Copy a string to the clipboard */
+export default function copyString(str: string) {
```
Comment (kyleshay) [inline src/app/utils/util.ts]: Hmm, I see we removed the copy-string.ts file, wasn't able to find a reference to it being used though.

File: src/app/dim-api/reducer.test.ts [hunk 2]
```diff
@@ -0,0 +1,564 @@
+import { DimApiState, initialState, dimApi } from './reducer';
```
Comment (kyleshay) [inline src/app/dim-api/reducer.test.ts]: Super cool! :0

File: src/app/storage/sync.service.ts [hunk 3]
```diff
@@ -68,18 +71,28 @@ const adapters: StorageAdapter[] = [new IndexedDBStorage(), GoogleDriveStorageAd
...
-      this.get(true).then(initSettings);
-    });
+      GoogleDriveStorageAdapter.signIn$.subscribe(() => {
+        // Force refresh data
+        console.log('GDrive sign in, refreshing data');
+        this.get(true).then(initSettings);
+      });
+    } else {
+      const apiScript = document.createElement('script');
+      apiScript.setAttribute('src', 'https://apis.google.com/js/api.js?callback=googleApiInit');
```
Comment (kyleshay) [inline src/app/storage/sync.service.ts]: I had some trouble finding the `googleApiInit` callback - I like that we're lazy loading the gapi now for users who haven't opted in yet!

Comment (kyleshay) [reply src/app/storage/sync.service.ts]: Ah, I see that `window.gapi_onload` should be called when this script loads to retrigger  SyncService.init() Maybe the callback isn't relevant.

File: src/Index.tsx [hunk 4]
```diff
@@ -50,7 +50,10 @@ if ($featureFlags.vendorEngrams) {
   saveVendorDropsToIndexedDB();
 }
 updateCSSVariables();
-store.dispatch(loadGlobalSettings());
+
+if ($featureFlags.dimApi) {
```
Comment (kyleshay) [inline src/Index.tsx]: I see `SyncService.init()` is called below (~L61) - should we consider moving it to the else block of this statement so we don't load the gapi for users who don't need it?

Comment (kyleshay) [reply src/Index.tsx]: Or if not the else statement, in the init() function of the SyncService, to return early if the dimAPI has loaded the config previously?

Comment (kyleshay) [reply src/Index.tsx]: Daaah, I see we're continuing to sync to google drive while we're in the beta. Sounds good!

File: src/app/utils/util.ts (hunk 1 above)
Comment (bhollis) [reply src/app/utils/util.ts]: It wasn't used anymore. I could remove it but I suspect we'll want it again soon.

File: src/app/storage/sync.service.ts (hunk 3 above)
Comment (bhollis) [reply src/app/storage/sync.service.ts]: Yeah I should go back through this - or maybe even remove it from this PR. My intention is to make it so we don't load gdrive unless people want to use it, but I backed off from a more aggressive cutover so we're always trying it now. I should at least make it predicated on whether you have Google Drive enabled.

File: src/Index.tsx (hunk 4 above)
Comment (bhollis) [reply src/Index.tsx]: Yeah originally I'd wanted it to be a bit more aggressive but backed off for now.

**Comment by bhollis** [comment]
Here goes nothin'

**Pull request merged**